File: SolutionCodes/Q18.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:

        # 16 ms solution
        if len(magazine) < len(ransomNote):
            return False

        for x in set(ransomNote):
            if ransomNote.count(x) > magazine.count(x):
                return False
        return True

        # 36 ms solution
        # for i in ransomNote:
        #     if i in magazine:
        #         magazine = magazine.replace(i,"", 1)
        #     else:
        #         return False
        # return True


# # sample 14000 KB submission
# class Solution:
#     def canConstruct(self, ransomNote: str, magazine: str) -> bool:
#         frequency ={}
#         for i in magazine:
#             if i not in frequency:
#                 frequency[i ] =0
#             frequency[i ]+ =1
#
#         for j in ransomNote:
#             if j not in frequency:
#                 return False
#             frequency[j ]- =1
#             if frequency[j ] <0:
#                 return False
#         return True

File: SolutionCodes/Q18.cpp
// 383. Ransom Note

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        int set[26]={};
        for(int i =0; i< magazine.size();++i)
            ++set[magazine[i]-'a'];

        int ransom[26]={};
        for(int j=0;j<ransomNote.size();++j)
            ++ransom[ransomNote[j]-'a'];

        for(int k = 0;k<26;++k)
            if(set[k]<ransom[k])
                return false;
        return true;
    }
};


/*
//19ms solution
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> map(26);
        for(int i=0;i<magazine.size();++i)
            ++map[magazine[i]];

        for(int j=0;j<ransomNote.size();++j)
        {   if(--map[ransomNote[j]]<0)
                return false;
        }
        return true;
    }
};

*/